tests/common/dualtor/tunnel_traffic_utils.py

```diff
@@ -0,0 +1,131 @@
+            """Find the ipv4 Loopback0 address."""
+            for addr in config_facts["LOOPBACK_INTERFACE"]["Loopback0"]:
+                if isinstance(ipaddress.ip_network(addr), ipaddress.IPv4Network):
+                    return addr.split("/")[0]
+
+        @staticmethod
+        def _build_tunnel_packet(out_src_ip, out_dst_ip):
+            """Build the expected tunnel packet."""
+            exp_pkt = testutils.simple_ip_packet(
+                ip_src=out_src_ip,
+                ip_dst=out_dst_ip,
+                pktlen=20
+            )
+            exp_pkt = mask.Mask(exp_pkt)
+            exp_pkt.set_do_not_care_scapy(Ether, "dst")
+            exp_pkt.set_do_not_care_scapy(Ether, "src")
+            exp_pkt.set_do_not_care_scapy(IP, "ihl")
+            exp_pkt.set_do_not_care_scapy(IP, "tos")
+            exp_pkt.set_do_not_care_scapy(IP, "len")
+            exp_pkt.set_do_not_care_scapy(IP, "id")
+            exp_pkt.set_do_not_care_scapy(IP, "flags")
+            exp_pkt.set_do_not_care_scapy(IP, "frag")
+            exp_pkt.set_do_not_care_scapy(IP, "ttl")
+            exp_pkt.set_do_not_care_scapy(IP, "proto")
+            exp_pkt.set_do_not_care_scapy(IP, "chksum")
+            exp_pkt.set_ignore_extra_bytes()
+            return exp_pkt
+
+        @staticmethod
+        def _dump_show_str(packet):
+            """Dump packet show output to string."""
+            _stdout, sys.stdout = sys.stdout, BytesIO()
+            try:
+                packet.show()
+                return sys.stdout.getvalue()
+            finally:
+                sys.stdout = _stdout
+
+        @staticmethod
+        def _check_ttl(packet):
+            """Check ttl field in the packet."""
+            outer_ttl, inner_ttl = packet[IP].ttl, packet[IP].payload[IP].ttl
+            logging.debug("Outer packet TTL: %s, inner packet TTL: %s", outer_ttl, inner_ttl)
+            if outer_ttl != inner_ttl:
+                raise ValueError("Outer packet TTL not same as inner packet TTL.")
+
+        @staticmethod
+        def _check_tos(packet):
+            """Check ToS field in the packet."""
+
+            def _disassemble_ip_tos(tos):
+                return tos >> 2, tos & 0x3
+
+            outer_tos, inner_tos = packet[IP].tos, packet[IP].payload[IP].tos
+            outer_dscp, outer_ecn = _disassemble_ip_tos(outer_tos)
+            inner_dscp, inner_ecn = _disassemble_ip_tos(inner_tos)
+            logging.debug("Outer packet DSCP: {0:06b}, inner packet DSCP: {1:06b}".format(outer_dscp, inner_dscp))
+            logging.debug("Outer packet ECN: {0:02b}, inner packet ECN: {0:02b}".format(outer_ecn, inner_ecn))
+            if outer_dscp != inner_ecn:
+                raise ValueError("Outer packet DSCP not same as inner packet DSCP.")
+            if outer_ecn != inner_ecn:
+                raise ValueError("Outer packet ECN not same as inner packet ECN")
+
+        def __init__(self, active_tor, standby_tor):
```

Comment:
In this util's design, it is testcase responsibility to send the packet. How about we add packet creation in the `__init__` function here? That will take some load from the testcases. If you agree that will be better, we can set some default values for packet attributes. For example, I think, these will mostly remain same for every testcase - `eth_dst, eth_src, ip_dscp, ip_ttl and ip_ecn`. The `ip_src and ip_dst` can be sent as an arg to `__init__`.

Reply:
I am also tempted to ask if we can include the ` testutils.send` also in this util itself. My idea is to keep testcases very light. If possible.

Reply:
I generally agree with this sentiment, but I think an issue here will be that we plan to use this util to verify tunnel traffic in end-to-end scenarios (e.g. using this alongside the dataplane utils) which may have some conflicts with the packets generated elsewhere.

Reply:
@theasianpianist, @vaibhavhd  Since some of the issues that depend on the tunnel traffic utility will generate traffic, I keep the code here simple as possible to only cover the traffic verification.